src/main/java/org/apache/commons/io/file/PathFence.java

```diff
@@ -0,0 +1,161 @@
+ * </pre>
+ * <p>
+ * To use the current directory as the path fence:
+ * </p>
+ * <pre>
+ * PathFence fence = PathFence.builder().setRoots(PathUtils.current()).get();
+ * </pre>
+ *
+ * @since 2.21.0
+ */
+// Cannot implement both UnaryOperator<Path> and Function<String, Path>, so don't pick one over the other
+public final class PathFence {
+
+    /**
+     * Builds {@link PathFence} instances.
+     */
+    public static final class Builder implements Supplier<PathFence> {
+
+        /** The empty Path array. */
+        private static final Path[] EMPTY = {};
+
+        /**
+         * A fence is made of root Paths.
+         */
+        private Path[] roots = EMPTY;
+
+        /**
+         * Constructs a new instance.
+         */
+        private Builder() {
+            // empty
+        }
+
+        @Override
+        public PathFence get() {
+            return new PathFence(this);
+        }
+
+        /**
+         * Sets the paths that delineate this fence.
+         *
+         * @param roots the paths that delineate this fence.
+         * @return {@code this} instance.
+         */
+        Builder setRoots(final Path... roots) {
+            this.roots = roots != null ? roots.clone() : EMPTY;
+            return this;
+        }
+    }
+
+    /**
+     * Creates a new builder.
+     *
+     * @return a new builder.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * A fence is made of Paths guarding Path resolution.
+     */
+    private final List<Path> roots;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param builder A builder.
+     */
+    private PathFence(final Builder builder) {
+        this.roots = Arrays.stream(builder.roots).map(Path::toAbsolutePath).collect(Collectors.toList());
+    }
+
+    /**
+     * Checks that that a Path resolves within our fence.
+     *
+     * @param path The path to test.
+     * @return The given path.
+     * @throws IllegalArgumentException Thrown if the file name is not without our fence.
+     */
+    // Cannot implement both UnaryOperator<Path> and Function<String, Path>
+    public Path apply(final Path path) {
+        return getPath(path.toString(), path);
+    }
+
+    /**
+     * Gets a Path for the given file name, checking that it resolves within our fence.
+     *
+     * @param fileName the file name to resolve.
+     * @return a fenced Path.
+     * @throws IllegalArgumentException Thrown if the file name is not without our fence.
+     */
+    // Cannot implement both UnaryOperator<Path> and Function<String, Path>
+    public Path apply(final String fileName) {
+        return getPath(fileName, Paths.get(fileName));
+    }
+
+    private Path getPath(final String fileName, final Path path) {
+        if (roots.isEmpty()) {
+            return path;
+        }
+        final Path pathAbs = path.normalize().toAbsolutePath();
```

Comment:
```suggestion
        final Path pathAbs = path.toAbsolutePath().normalize();
```

If you call it in the reverse order, `normalize()` will not remove leading `..` segments. For example on UNIX:

- if `path` equals `../../etc/passwd`,
- then `path.normalize()` equals `../../etc/passwd`,
- and `path.normalize().toAbsolutePath()` equals `/home/piotr/../../etc/passwd`.